engines/groovie/roq.cpp

```diff
@@ -124,31 +196,120 @@ uint16 ROQPlayer::loadInternal() {
 	}
 }
 
+
+// Calculate the overlapping area for the rendered frame to the visible frame. The game can use an origin to
+// place the rendered image at different places.
+void ROQPlayer::calcStartStop(int &start, int &stop, int origin, int length) {
+	if (origin >= 0) {
+		start = origin;
+		stop = length;
+	} else {
+		start = 0;
+		stop = length + origin;
+	}
+}
+
 void ROQPlayer::buildShowBuf() {
-	if (_alpha)
-		_fg->copyFrom(*_bg);
-
-	for (int line = 0; line < _bg->h; line++) {
-		uint32 *out = _alpha ? (uint32 *)_fg->getBasePtr(0, line) : (uint32 *)_bg->getBasePtr(0, line);
-		uint32 *in = (uint32 *)_currBuf->getBasePtr(0, line / _scaleY);
-
-		for (int x = 0; x < _bg->w; x++) {
-			// Copy a pixel, checking the alpha channel first
-			if (_alpha && !(*in & 0xFF))
-				out++;
-			else if (_fg->h == 480 && *in == _vm->_pixelFormat.RGBToColor(255, 255, 255))
-				// Handle transparency in Gamepad videos
-				// TODO: For now, we detect these videos by checking for full screen
-				out++;
-			else
-				*out++ = *in;
+	// Calculate screen offset for normal / fullscreen videos and images
+	int screenOffset = 0;
+	if (_screen->h != 480) {
+		screenOffset = 80;
+	}
+
+	// Restore the background by data from the foreground. Only restore the area which was overwritten during the last frame
+	// Therefore we have the _restoreArea which reduces the area for restoring. We also use the _prevBuf to only overwrite the
+	// Pixels which have been written during the last frame. This means _restoreArea is just an optimization.
+	if (_alpha) {
+		if (!_restoreArea->isEmpty()) {
+			int width = _restoreArea->right - _restoreArea->left;
+			Graphics::Surface *screen = _vm->_system->lockScreen();
+			for (int line = _restoreArea->top; line < _restoreArea->bottom; line++) {
+				byte *dst = (byte *)screen->getBasePtr(_restoreArea->left, line + screenOffset);
+				byte *src = (byte *)_bg->getBasePtr(_restoreArea->left, line);
+				byte *prv = (byte *)_prevBuf->getBasePtr((_restoreArea->left - _origX) / _scaleX, (line - _origY) / _scaleY);
+				byte *ovr = (byte *)_overBuf->getBasePtr(_restoreArea->left, line);
+				for (int i = 0; i < width; i++) {
```

Comment:
This potentially will be a CPU hog at slow machines. Is there a way to keep all your buffers in the same format, so you could do simple byte copying here?